### PR TITLE
[WIP] Add nix flake to provide development shell

### PR DESCRIPTION
**Pull request by RCoeurjoly**
Tested by executing:
`make install-opam-deps`
and 
`make universe`
and all the jobs complete successfully

### CODE REVIEW

**Comment by Alizter** [comment]
Did you mean to point to your own branch of VST? They don't have dune support right?

**Comment by RCoeurjoly** [comment]
Sorry, this PR only deals with the nix flake. I pushed to my fork and github decided to add that commit to this PR.
About VST, let´s continue the conversation in Zulip.

**Comment by ejgallego** [comment]
> make install-opam-deps

Actually you could also install the deps from Nix directly, but we can do that later on.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -45,3 +45,20 @@ jobs:
 
       - name: 🌌 Coq Universe
         run: opam exec -- make universe DUNEOPT=""
+  nix:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: cachix/install-nix-action@v17
+
+    - name: Enter development shell
```
Comment (Alizter) [inline .github/workflows/ci.yml:54]:
```suggestion
    - name: 🐚 Enter development shell
```

**Comment by ejgallego** [comment]
Nix CI is broken.

**Comment by RCoeurjoly** [comment]
I see that nix has reverted the support for submodules, and is working on supporting them again (https://github.com/NixOS/nix/pull/5497).
So maybe only provide the shell and latter do the CI?

**Comment by ejgallego** [comment]
@RCoeurjoly how's the CI error related to submodules? I see:
```
make: opam: Command not found
```
so that def means something is wrong.

**Comment by RCoeurjoly** [comment]
After some tests, I see the following:

- Using the devShell (nix develop) to build the packages in CI is unfeasible because there is no way to execute something programmatically in that shell
- nix build doesn´t support submodules
- nix shell can take an argument with the commands to execute, however
  - You would need to duplicate the development packages like this:
`  nix shell nixpkgs#gmp nixpkgs#dune_3 nixpkgs#ocaml nixpkgs#opam -c bash script.sh`
  - you cannot execute a complex command with &&, you would need to put it in a script file.

No option is feasible in my opinion.

**Comment by ejgallego** [comment]
@RCoeurjoly that's a bit strange, I'm no Nix expert but in Coq we have CI for this kind of stuff without a problem.

It seems pretty strange that a sequence of commands that works locally would not work in the CI (i.e. option 1)

**Comment by Alizter** [comment]
There are nix workflows for github that we could use. These automatically setup the flake in the repo and give you a development shell.

**Pull request merged**